### PR TITLE
[ELF] Align the end of PT_GNU_RELRO associated PT_LOAD to a common-page-size boundary

### LINKED ISSUE #57618: lld produces non-runnable ELF if common-page-size and max-page-size options are used
Hi,

Linking ELF file with lld and `-Wl,-zcommon-page-size=2097152 -Wl,-zmax-page-size=2097152` produces ELF which crashes on start with error:
```
/opt/php/bin/php: error while loading shared libraries: cannot apply additional memory protection after relocation: Cannot allocate memory
```

Linking with `-Wl,-zmax-page-size=2097152` only produces ELF which has align 0x200000 in readelf's output but no sections are actually aligned to this value. ld works fine in both cases.

I've encountered this error in a private codebase but since producing minimal testcase is not trivial and there was [exact same issue in PHP](https://bugs.php.net/bug.php?id=79092&edit=3), I decided to refresh their reproducer:

```dockerfile
FROM ubuntu:22.04

RUN apt-get update && \
    apt-get install -y \
    autoconf curl gnupg2 lsb-release make \
    software-properties-common wget

WORKDIR /root
RUN wget https://apt.llvm.org/llvm.sh && \
    chmod +x llvm.sh && \
    ./llvm.sh 15

ARG PHP_VERSION="php-8.1.9"
RUN curl -sL https://www.php.net/distributions/${PHP_VERSION}.tar.xz | tar xJv
WORKDIR /root/${PHP_VERSION}

# uncomment to set common-page-size and produce invalid ELF
#RUN sed -i 's/EXTRA_LDFLAGS_PROGRAM -Wl,-zmax-page-size=2097152/EXTRA_LDFLAGS_PROGRAM -Wl,-zcommon-page-size=2097152 -Wl,-zmax-page-size=2097152/g' ./configure.ac && ./buildconf --force

ARG CFLAGS="-Oz"
ARG LDFLAGS="-fuse-ld=lld-15"
RUN CC="clang-15" CFLAGS="${CFLAGS}" LDFLAGS="${LDFLAGS}" ./configure --prefix=/opt/php --disable-cgi --disable-all && \
    make -j $(nproc) && \
    make install

RUN echo "<?php phpinfo();" | /opt/php/bin/php ; readelf -WlS /opt/php/bin/php
```

### LINKED ISSUE #65002: lld doesn't protect the last RELRO page on Asahi Linux
I installed Asahi Linux on my M1 Mac Mini. Note that Asahi Linux is a 16 KiB page system.

```
#include <stdio.h>

extern char _GLOBAL_OFFSET_TABLE_;

int main() {
  printf("Hello world\n");
  _GLOBAL_OFFSET_TABLE_ = 3;
}
```

If I run the above program, it should crash because it writes to `.got` and `.got` should be protected by GNU_RELRO. However, it doesn't actually crash.

```
$ clang -fuse-ld=/home/ruiu/llvm-project/build/bin/ld.lld -o hello hello.c
$ ./hello
Hello world
```

I believe it is because RELRO does not end at a page boundary and therefore the last page in the RELRO isn't protected. As you can see, the RELRO in the executable ends at 0x21000, not at 0x30000.

```
$ readelf --segments hello

Elf file type is DYN (Position-Independent Executable file)
Entry point 0x10780
There are 11 program headers, starting at offset 64

Program Headers:
  Type           Offset   VirtAddr           PhysAddr           FileSiz  MemSiz   Flg Align
  PHDR           0x000040 0x0000000000000040 0x0000000000000040 0x000268 0x000268 R   0x8
  INTERP         0x0002a8 0x00000000000002a8 0x00000000000002a8 0x00001b 0x00001b R   0x1
      [Requesting program interpreter: /lib/ld-linux-aarch64.so.1]
  LOAD           0x000000 0x0000000000000000 0x0000000000000000 0x000750 0x000750 R   0x10000
  LOAD           0x000780 0x0000000000010780 0x0000000000010780 0x0001d0 0x0001d0 R E 0x10000
  LOAD           0x000950 0x0000000000020950 0x0000000000020950 0x0001f0 0x0001f0 RW  0x10000
  LOAD           0x000b40 0x0000000000030b40 0x0000000000030b40 0x000048 0x000049 RW  0x10000
  DYNAMIC        0x000960 0x0000000000020960 0x0000000000020960 0x0001b0 0x0001b0 RW  0x8
  GNU_RELRO      0x000950 0x0000000000020950 0x0000000000020950 0x0001f0 0x0006b0 R   0x1
  GNU_EH_FRAME   0x000644 0x0000000000000644 0x0000000000000644 0x00003c 0x00003c R   0x4
  GNU_STACK      0x000000 0x0000000000000000 0x0000000000000000 0x000000 0x000000 RW  0
  NOTE           0x0002c4 0x00000000000002c4 0x00000000000002c4 0x000038 0x000038 R   0x4
```

### PR DESCRIPTION
**Pull request by MaskRay**
Close #57618: currently we align the end of PT_GNU_RELRO to a common-page-size
boundary, but do not align the end of the associated PT_LOAD. This is benign
when runtime_page_size >= common-page-size.

However, when runtime_page_size < common-page-size, it is possible that
`alignUp(end(PT_LOAD), page_size) < alignDown(end(PT_GNU_RELRO), page_size)`.
In this case, rtld's mprotect call for PT_GNU_RELRO will apply to unmapped
regions and lead to an error, e.g.

```
error while loading shared libraries: cannot apply additional memory protection after relocation: Cannot allocate memory
```

To fix the issue, add a padding section .relro_padding like mold, which
is contained in the PT_GNU_RELRO segment and the associated PT_LOAD
segment. The section also prevents strip from corrupting PT_LOAD program
headers.

.relro_padding has the largest `sortRank` among RELRO sections.
Therefore, it is naturally placed at the end of `PT_GNU_RELRO` segment
in the absence of `PHDRS`/`SECTIONS` commands.

In the presence of `SECTIONS` commands, we place .relro_padding
immediately before a symbol assignment using DATA_SEGMENT_RELRO_END (see
also https://reviews.llvm.org/D124656), if present.
DATA_SEGMENT_RELRO_END is changed to align to max-page-size instead of common-page-size.

Some edge cases worth mentioning:

* ppc64-toc-addis-nop.s: when PHDRS is present, do not append .relro_padding
* avoid-empty-program-headers.s: when the only RELRO section is .tbss,
  it is not part of PT_LOAD segment, therefore we do not append .relro_padding.

---

Close #65002: GNU ld from 2.39 onwards aligns the end of PT_GNU_RELRO to a
max-page-size boundary (https://sourceware.org/PR28824) so that the last page is
protected even if runtime_page_size > common-page-size.

In my opinion, losing protection for the last page when the runtime page size is
larger than common-page-size is not really an issue. Double mapping a page of up
to max-common-page for the protection could cause undesired VM waste. Internally
we had users complaining about 2MiB max-page-size applying to shared objects.

Therefore, the end of .relro_padding is padded to a common-page-size
boundary. Users who are really anxious can set common-page-size to match
their runtime page size.

---

17 tests need updating as there are lots of change detectors.

### CODE REVIEW

**Review by smithp35 — COMMENTED** [review]
I've not had a chance to go through the code changes in detail yet. 

Can you help me out a bit with what is causing the problem? Is it just that strip is corrupting the output, or is that just one of the problems?

The description says "system-page-size < common-page-size". 

What does system-page-size refer to? Do you mean max-page-size? Or is it something like the page size used by the OS? For example: AArch64 max-page-size = 64Kib common-page-size = 4 KiB. An OS may set system-page-size to 4KiB but if I link with max-page-size and common-page-size set to 64 KiB then system-page-size < common-page (and max-page-size).

I mention that because we currently don't permit common-page-size to be larger than max-page-size.
```
getCommonPageSize()
...
  // commonPageSize can't be larger than maxPageSize.                                                                                                               if (val > config->maxPageSize)
    val = config->maxPageSize;
```

In the description
> in the absence of PHDRS/SECTIONS: ensure the end of PT_GNU_RELRO associated PT_LOAD to a common-page-size boundary

Looks like the sentence is missing a few words on alignment. Perhaps: 
> ensure the end of the PT_LOAD program loader associated with the PT_GNU_RELRO segment is aligned to a common-page-size boundary.

**Comment by MaskRay** [comment]
> I've not had a chance to go through the code changes in detail yet.
> 
> Can you help me out a bit with what is causing the problem? Is it just that strip is corrupting the output, or is that just one of the problems?
> 
> The description says "system-page-size < common-page-size".
> 
> What does system-page-size refer to? Do you mean max-page-size? Or is it something like the page size used by the OS? For example: AArch64 max-page-size = 64Kib common-page-size = 4 KiB. An OS may set system-page-size to 4KiB but if I link with max-page-size and common-page-size set to 64 KiB then system-page-size < common-page (and max-page-size).
> 
> I mention that because we currently don't permit common-page-size to be larger than max-page-size.
> 
> ```
> getCommonPageSize()
> ...
>   // commonPageSize can't be larger than maxPageSize.                                                                                                               if (val > config->maxPageSize)
>     val = config->maxPageSize;
> ```

Thanks for the comments. I used system-page-size to mean the OS's page-size. I have renamed it to "runtime page-size" to improve clarity.

Updated the description of the PR:

    Close #57618: currently we align the end of PT_GNU_RELRO to a common-page-size
    boundary, but do not align the end of the associated PT_LOAD, i.e.
    when runtime_page_size < common-page-size, it is possible that `alignUp(end(PT_LOAD), page_size) > alignDown(end(PT_GNU_RELRO), page_size)`.
    rtld calling `mprotect` on unmapped regions gives an error, e.g.

    ```
    error while loading shared libraries: cannot apply additional memory protection after relocation: Cannot allocate memory
    ```

The issue was why [[ELF] Align the end of PT_GNU_RELRO to max-page-size instead of common-page-size](https://reviews.llvm.org/D125410) was incorrect.

> In the description
> 
> > in the absence of PHDRS/SECTIONS: ensure the end of PT_GNU_RELRO associated PT_LOAD to a common-page-size boundary
> 
> Looks like the sentence is missing a few words on alignment. Perhaps:
> 
> > ensure the end of the PT_LOAD program loader associated with the PT_GNU_RELRO segment is aligned to a common-page-size boundary.

Updated. Thanks for the suggestion.

**Comment by rui314** [comment]
Maybe you want to leave a comment somewhere in the code about what is `.relro_padding`? `.relro_padding` is there so that all pages in data segments are covered by at least one section. My understanding is that the section (or the section table in general) is not necessary for runtime because the runtime only cares about the program header, so you can technically leave the end of the data segment unused, but the `strip` tool would truncate a segment if the end of the segment is not covered by any section.

**Comment by MaskRay** [comment]
Improve comments to RelroPaddingSection. Thanks to rui314

File: lld/ELF/Writer.cpp [hunk 1]
```diff
@@ -920,7 +929,9 @@ static unsigned getSectionRank(const OutputSection &osec) {
     // TLS sections directly before the other RELRO sections.
     if (!(osec.flags & SHF_TLS))
       rank |= RF_NOT_TLS;
-    if (!isRelroSection(&osec))
+    if (isRelroSection(&osec))
+      osec.relro = true;
```
Comment (smithp35) [inline lld/ELF/Writer.cpp:933]: IIUC this was previously used for sections between DATA_SEGMENT_ALIGN and DATA_SEGMENT_RELRO_END. With it being cleared if there is no DATA_SEGMENT_RELRO_END. Wouldn't this unconditionally add it back in again even if there was no DATA_SEGMENT_RELRO_END?

File: lld/ELF/LinkerScript.cpp [hunk 2]
```diff
@@ -1196,9 +1202,17 @@ void LinkerScript::adjustOutputSections() {
...
+    // Note: when .tbss is the only RELRO section, it's difficult to pick a
+    // suitable padding size (see computeFilOffset). For simplicity, don't
+    // retain .relro_padding in this case.
+    if (in.relroPadding && in.relroPadding->getParent() == sec && !seenRelro)
+      discardable = true;
     if (discardable) {
       sec->markDead();
       cmd = nullptr;
+    } else {
+      seenRelro |= sec->relro && sec->type != SHT_NOBITS;
```
Comment (smithp35) [inline lld/ELF/LinkerScript.cpp:1215]: Would it be worth restricting this to sec->type != SHT_NOBITS && (sec->flags & SHF_TLS)

File: lld/ELF/LinkerScript.cpp [hunk 3]
```diff
@@ -1196,9 +1202,17 @@ void LinkerScript::adjustOutputSections() {
     if (sec->sectionIndex != UINT32_MAX)
       maybePropagatePhdrs(*sec, defPhdrs);
 
+    // Discard .relro_padding if we have not seen one non-NOBITS RELRO section.
+    // Note: when .tbss is the only RELRO section, it's difficult to pick a
```
Comment (smithp35) [inline lld/ELF/LinkerScript.cpp:1206]: I can see that a .tbss only (which would have no impact on the PT_LOAD) would be best excluded. It looks like the code would also exclude .bss.rel.ro only. Although I'm not entirely sure if any tool actually generates those.

**Review by smithp35 — COMMENTED** [review]
Thanks for updating the description, it is clearer what the problem is now.

Is the concern about the padding section that there may be no pattern, or orphan placement might put it in the wrong place? For example in a linker script with DATA_SEGMENT_RELRO_END I could see it being worth inserting the padding section in that case as there would be a clear insertion point.

I agree that most people using linker scripts are embedded users, although another use case is using ld --verbose and editing it then using it with lld. This would have DATA_SEGMENT_RELRO_END.

Are there sufficient differences from GNU ld to make it worth updating the linker script differences doc https://lld.llvm.org/ELF/linker_script.html

File: lld/ELF/Writer.cpp (hunk 1 above)
Comment (MaskRay) [reply lld/ELF/Writer.cpp:933]: `ScriptParser::readSections` clears `relro` if there is no `DATA_SEGMENT_RELRO_END`.

After `getSectionRank`, `relro` has changed its meaning: whether the section is RELRO, to be checked in `seenRelro |= sec->relro && sec->type != SHT_NOBITS;`.

In the absence of `DATA_SEGMENT_RELRO_END`, this code is still correct.

File: lld/ELF/LinkerScript.cpp (hunk 2 above)
Comment (MaskRay) [reply lld/ELF/LinkerScript.cpp]: Looks good. Added a new test to `relocation-copy-relro.s` where .bss.rel.ro is the only RELRO section and we still append the padding.

**Comment by MaskRay** [comment]
> Thanks for updating the description, it is clearer what the problem is now.

Thanks for the suggestion!

> Is the concern about the padding section that there may be no pattern, or orphan placement might put it in the wrong place? For example in a linker script with DATA_SEGMENT_RELRO_END I could see it being worth inserting the padding section in that case as there would be a clear insertion point.
>
> I agree that most people using linker scripts are embedded users, although another use case is using ld --verbose and editing it then using it with lld. This would have DATA_SEGMENT_RELRO_END.

Yes, I wanted to avoid dealing with orphan placement. It turns out that supporting `.relro_padding` in the presence of `DATA_SEGMENT_RELRO_END` isn't too hard. I have added the support in 8790f8f34841041828202a2dc7bec66119da23d4
 
> Are there sufficient differences from GNU ld to make it worth updating the linker script differences doc https://lld.llvm.org/ELF/linker_script.html

Added some documentation.

**Review by smithp35 — APPROVED** [review]
Thanks for the updates and clarification. Looks good to me now.

**Comment by jplehr** [comment]
This seems to have broken the AMDGPU OpenMP staging buildbot https://lab.llvm.org/staging/#/builders/247/builds/6482

File: lld/test/ELF/linkerscript/data-segment-relro-ppc64.test [hunk 4]
```diff
@@ -0,0 +1,79 @@
+# REQUIRES: x86
```
Comment (dyung) [inline lld/test/ELF/linkerscript/data-segment-relro-ppc64.test:1]: Should this be ppc? It is causing failures on bots that don't build PPC such as the PS bots:
https://lab.llvm.org/buildbot/#/builders/139/builds/49724
https://lab.llvm.org/buildbot/#/builders/216/builds/27302

Comment (dyung) [reply lld/test/ELF/linkerscript/data-segment-relro-ppc64.test:1]: I have made a change to the REQUIRES line to be `ppc` instead in 72bbac4.

**Comment by gulfemsavrun** [comment]
We started seeing a test failure:
```
FAIL: lld :: ELF/linkerscript/data-segment-relro-ppc64.test (1203 of 2824)
******************** TEST 'lld :: ELF/linkerscript/data-segment-relro-ppc64.test' FAILED ********************
Script:
--
: 'RUN: at line 2';   rm -rf /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp && split-file /b/s/w/ir/x/w/llvm-llvm-project/lld/test/ELF/linkerscript/data-segment-relro-ppc64.test /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp
: 'RUN: at line 3';   /b/s/w/ir/x/w/llvm_build/bin/llvm-mc -filetype=obj -triple=powerpc64le /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/a.s -o /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/a.o
: 'RUN: at line 4';   /b/s/w/ir/x/w/llvm_build/bin/llvm-mc -filetype=obj -triple=powerpc64le /b/s/w/ir/x/w/llvm-llvm-project/lld/test/ELF/linkerscript/Inputs/shared.s -o /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/b.o
: 'RUN: at line 5';   /b/s/w/ir/x/w/llvm_build/bin/ld.lld -shared -soname=b /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/b.o -o /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/b.so
: 'RUN: at line 7';   /b/s/w/ir/x/w/llvm_build/bin/ld.lld -z max-page-size=65536 -z norelro /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/a.o /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/b.so -T /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/1.t -o /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/a1
: 'RUN: at line 10';   /b/s/w/ir/x/w/llvm_build/bin/ld.lld -z max-page-size=65536 -z relro /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/a.o /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/b.so -T /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/1.t -o /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/a2
: 'RUN: at line 11';   /b/s/w/ir/x/w/llvm_build/bin/llvm-readelf -S -l /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/a2 | /b/s/w/ir/x/w/llvm_build/bin/FileCheck /b/s/w/ir/x/w/llvm-llvm-project/lld/test/ELF/linkerscript/data-segment-relro-ppc64.test --check-prefixes=CHECK2
--
Exit Code: 1

Command Output (stderr):
--
```

https://luci-milo.appspot.com/ui/p/fuchsia/builders/toolchain.ci/clang-linux-x64/b8769957246457079665/overview

**Comment by MaskRay** [comment]
> We started seeing a test failure:
> 
> ```
> FAIL: lld :: ELF/linkerscript/data-segment-relro-ppc64.test (1203 of 2824)
> ******************** TEST 'lld :: ELF/linkerscript/data-segment-relro-ppc64.test' FAILED ********************
> Script:
> --
> : 'RUN: at line 2';   rm -rf /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp && split-file /b/s/w/ir/x/w/llvm-llvm-project/lld/test/ELF/linkerscript/data-segment-relro-ppc64.test /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp
> : 'RUN: at line 3';   /b/s/w/ir/x/w/llvm_build/bin/llvm-mc -filetype=obj -triple=powerpc64le /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/a.s -o /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/a.o
> : 'RUN: at line 4';   /b/s/w/ir/x/w/llvm_build/bin/llvm-mc -filetype=obj -triple=powerpc64le /b/s/w/ir/x/w/llvm-llvm-project/lld/test/ELF/linkerscript/Inputs/shared.s -o /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/b.o
> : 'RUN: at line 5';   /b/s/w/ir/x/w/llvm_build/bin/ld.lld -shared -soname=b /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/b.o -o /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/b.so
> : 'RUN: at line 7';   /b/s/w/ir/x/w/llvm_build/bin/ld.lld -z max-page-size=65536 -z norelro /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/a.o /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/b.so -T /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/1.t -o /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/a1
> : 'RUN: at line 10';   /b/s/w/ir/x/w/llvm_build/bin/ld.lld -z max-page-size=65536 -z relro /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/a.o /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/b.so -T /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/1.t -o /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/a2
> : 'RUN: at line 11';   /b/s/w/ir/x/w/llvm_build/bin/llvm-readelf -S -l /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/a2 | /b/s/w/ir/x/w/llvm_build/bin/FileCheck /b/s/w/ir/x/w/llvm-llvm-project/lld/test/ELF/linkerscript/data-segment-relro-ppc64.test --check-prefixes=CHECK2
> --
> Exit Code: 1
> 
> Command Output (stderr):
> --
> ```
> 
> https://luci-milo.appspot.com/ui/p/fuchsia/builders/toolchain.ci/clang-linux-x64/b8769957246457079665/overview

We started seeing a test failure:

```
FAIL: lld :: ELF/linkerscript/data-segment-relro-ppc64.test (1203 of 2824)
******************** TEST 'lld :: ELF/linkerscript/data-segment-relro-ppc64.test' FAILED ********************
Script:
--
: 'RUN: at line 2';   rm -rf /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp && split-file /b/s/w/ir/x/w/llvm-llvm-project/lld/test/ELF/linkerscript/data-segment-relro-ppc64.test /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp
: 'RUN: at line 3';   /b/s/w/ir/x/w/llvm_build/bin/llvm-mc -filetype=obj -triple=powerpc64le /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/a.s -o /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/a.o
: 'RUN: at line 4';   /b/s/w/ir/x/w/llvm_build/bin/llvm-mc -filetype=obj -triple=powerpc64le /b/s/w/ir/x/w/llvm-llvm-project/lld/test/ELF/linkerscript/Inputs/shared.s -o /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/b.o
: 'RUN: at line 5';   /b/s/w/ir/x/w/llvm_build/bin/ld.lld -shared -soname=b /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/b.o -o /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/b.so
: 'RUN: at line 7';   /b/s/w/ir/x/w/llvm_build/bin/ld.lld -z max-page-size=65536 -z norelro /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/a.o /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/b.so -T /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/1.t -o /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/a1
: 'RUN: at line 10';   /b/s/w/ir/x/w/llvm_build/bin/ld.lld -z max-page-size=65536 -z relro /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/a.o /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/b.so -T /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/1.t -o /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/a2
: 'RUN: at line 11';   /b/s/w/ir/x/w/llvm_build/bin/llvm-readelf -S -l /b/s/w/ir/x/w/llvm_build/tools/lld/test/ELF/linkerscript/Output/data-segment-relro-ppc64.test.tmp/a2 | /b/s/w/ir/x/w/llvm_build/bin/FileCheck /b/s/w/ir/x/w/llvm-llvm-project/lld/test/ELF/linkerscript/data-segment-relro-ppc64.test --check-prefixes=CHECK2
--
Exit Code: 1

Command Output (stderr):
--
```

Fixed by 72bbac47389918502f95423eaeb90733a6b8bb89
```
--- a/lld/test/ELF/linkerscript/data-segment-relro-ppc64.test
+++ b/lld/test/ELF/linkerscript/data-segment-relro-ppc64.test
@@ -1,4 +1,4 @@
-# REQUIRES: x86
+# REQUIRES: ppc
```

**Pull request merged**